### PR TITLE
Pool JsonSerializer instances

### LINKED ISSUE #3550: Pool JsonSerializer instances
See data here:

image

See comment from @JamesNK here: https://github.com/aspnet/Mvc/issues/3542#issuecomment-156200726

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3550

### CODE REVIEW

**Comment by pranavkm** [comment]
image

**Before**
image


**After**
image

**Comment by pranavkm** [comment]
cc @rynowak

**Comment by davidfowl** [comment]
Those results are pretty crazy

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 1]
```diff
@@ -4,47 +4,87 @@
...
+
+            if (objectPoolProvider == null)
+            {
+                throw new ArgumentNullException(nameof(objectPoolProvider));
             }
 
             _logger = logger;
             _serializerSettings = serializerSettings;
             _charPool = new JsonArrayPool<char>(charPool);
+            _jsonSerializerPool = objectPoolProvider.Create(new JsonSerializerObjectPolicy(_serializerSettings));
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: We should either make the `SerializerSettings` non-settable, or make this lazy.

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 2]
```diff
@@ -143,10 +182,11 @@ public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterCo
                     }
                     finally
                     {
-                        // Clean up the error handler in case CreateJsonSerializer() reuses a serializer
+                        // Clean up the error handler since CreateJsonSerializer() pools instances.
                         jsonSerializer.Error -= errorHandler;
                     }
 
+                    ReleaseJsonSerializer(jsonSerializer);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: Doesn't this need to be in the finally?

**Comment by pranavkm** [comment]
Updated

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: Yeah might as well, since we have it.

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**